File: Patterns/Command Pattern/CSharp/Command/Light/LightOffCommand.cs
namespace Command.Light
{
    public class LightOffCommand : ICommand
    {
        private Light light;
        private int preview;

        public LightOffCommand(Light light)
        {
            preview = Light.OFF;
            this.light = light;
        }

        public void Execute()
        {
            preview = light.State;
            light.Off();
        }

        public void Undo()
        {
            if (preview == Light.ON)
            {
                light.On();
            }
        }
    }
}

File: Patterns/Command Pattern/Cpp/Command/Light/LightOffCommand.cpp
#include "../ICommand.cpp"
#include "./Light.cpp"

class LightOffCommand : public ICommand {
private:
    Light* light;
    int preview;

public:
    LightOffCommand(Light* light) : light{light}, preview{light->OFF} {}

    void execute() override {
        preview = light->getState();
        light->off();
    }

    void undo() override {
        if (preview == light->ON) light->on();
    }

    string getCommandName() override {
        return "LightOffCommand";
    }
};